diff --git a/src/devices/machine/z8010.cpp b/src/devices/machine/z8010.cpp
index 5952df1c4c2..98417843d3e 100644
--- a/src/devices/machine/z8010.cpp
+++ b/src/devices/machine/z8010.cpp
@@ -37,8 +37,7 @@ DEFINE_DEVICE_TYPE(Z8010, z8010_device, "z8010", "Zilog Z8010 MMU")
 
 inline void z8010_device::INC_SAR()
 {
-	if (m_sar < (SDR_MAX_SEGMENTS-1))
-		++m_sar;
+	m_sar = (m_sar + 1) & (SDR_MAX_SEGMENTS - 1);
 }
 
 //-------------------------------------------------
@@ -63,10 +62,10 @@ inline void z8010_device::INC_DSC_SAR(const uint8_t max_dsc = DSC_SDR_ATTR)
 	{
 		++m_dsc;
 	}
-	else if (m_sar < (SDR_MAX_SEGMENTS-1))
+	else
 	{
 		m_dsc = DSC_SDR_BAH;
-		++m_sar;
+		INC_SAR();
 	}
 }
 
